frontend/app/routes/widget.tsx

```diff
@@ -0,0 +1,502 @@
+                          className="xl:w-[250px] xl:rounded-lg
+                                     w-full min-w-0 order-first xl:order-last"
+                          disabled={isLoadingScript}
+                          onClick={() => handleSave('script')}
+                        >
+                          <div className="flex items-center justify-center gap-xs">
+                            {isLoadingScript && (
+                              <SVGSpinner className="w-4 h-4" />
+                            )}
+                            <span>
+                              {isLoadingScript
+                                ? 'Saving...'
+                                : 'Save and generate script'}
+                            </span>
+                          </div>
+                        </ToolsPrimaryButton>
+                      </div>
+                    </div>
+                  </div>
+
+                  <div
+                    id="preview"
+                    className="w-full mx-auto xl:mx-0 xl:sticky xl:top-md xl:self-start xl:flex-shrink-0 xl:w-[504px] h-fit"
+                  >
+                    <BuilderBackground>
+                      <WidgetPreview />
+                    </BuilderBackground>
+                  </div>
+                </div>
+              </div>
+            </div>
+          </div>
+        </div>
+      </div>
+
+      {snap.modal?.type === 'script' && (
+        <div className="fixed inset-0 bg-slate-500 bg-opacity-75 transition-opacity z-50">
+          <div className="fixed inset-0 z-50 overflow-y-auto">
+            <div className="flex min-h-full items-center justify-center p-4 text-center">
+              <ScriptReadyModal
+                isOpen={true}
+                onClose={handleCloseModal}
+                scriptContent={toolActions.getScriptToDisplay()}
+                onCopy={() => console.log('Script copied to clipboard')}
+              />
+            </div>
+          </div>
+        </div>
+      )}
+
+      {snap.modal?.type === 'save-success' && (
+        <div className="fixed inset-0 bg-slate-500 bg-opacity-75 transition-opacity z-50">
+          <div className="fixed inset-0 z-50 overflow-y-auto">
+            <div className="flex min-h-full items-center justify-center p-4 text-center">
+              <SaveResultModal
+                isOpen={true}
+                onClose={handleCloseModal}
+                onDone={handleCloseModal}
+                message="Your edits have been saved"
+                isSuccess={true}
+              />
+            </div>
+          </div>
+        </div>
+      )}
+
+      {snap.modal?.type === 'save-error' && (
+        <div className="fixed inset-0 bg-slate-500 bg-opacity-75 transition-opacity z-50">
+          <div className="fixed inset-0 z-50 overflow-y-auto">
+            <div className="flex min-h-full items-center justify-center p-4 text-center">
+              <SaveResultModal
```

Comment:
Consider moving all these modals into a separate component. Then the `Widget` will be smaller and easier to scan.